src/repo/fs.rs

```diff
@@ -88,5 +88,51 @@ impl DataStore for FsDataStore {
     }
 }
 
+use fs2::FileExt;
+use std::fs::File;
+use std::path::Path;
+
+#[derive(Debug)]
+pub struct FsLock {
+    file: File,
+}
+
+impl Lock for FsLock {
+    fn new(path: &Path) -> std::io::Result<Self> {
+        use std::fs::OpenOptions;
+
+        let file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .open(path)
+            .unwrap();
```

Comment:
In general, under which conditions does this panic? The home is a read-only directory?

While in general this panicing for such an initialization error would be ok for *an application*, perhaps we should find a way to return the error, and only do this at the `Repo::init` time? That would probably mean taking in a PathBuf instead, holding on to it until we can open it as a file, adding an error type which would only be used internally (or be public), cast to `anyhow::Error` on the `UninitializedIpfs::start` https://github.com/rs-ipfs/rust-ipfs/blob/d3c84ee52e926c1c2370f09515db02d5d9dd584c/src/lib.rs#L335

Reply:
Jotting down some thoughts here: 

> In general, under which conditions does this panic? The home is a read-only directory?

Bad path, permissions, mentioned [here](https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#errors). This specific error path might also be relevant for the rest of the API (at least the `fs` impl). 

`try_lock_exclusive` errors are os errors, as an example this is [the code for unix](https://github.com/danburkert/fs2-rs/blob/master/src/unix.rs#L46-L48). We could also go the rustc route and block whilst waiting to acquire a lock though I'm thinking a nice `Another IPFS node is already running with this peer id` would be better.`

In order to lock the repo on `init` instead of `new`, I'm thinking the `Lock` struct will need a state key with an Enum variant (i.e. Lock with pathbuf in `repo::new`, actually create the Lock in `init`). Something like [this](https://github.com/rust-lang/cargo/blob/02205f5403d33d5d1666fc7b9f6fd321de511878/src/cargo/util/flock.rs#L13-L25)?